### PR TITLE
feat: add moduleDeps classes to the classpath

### PR DESCRIPTION
**Pull request by jhenahan**
# Motivation

In my code, I am building images with this plugin, but I've had trouble getting the resulting images to run due to `ClassDefNotFound` errors on startup. Eventually it clicked that these were classes from my `moduleDeps`, and sure enough `dive` showed `app/classes.jar/` was empty in the resulting images. I think this is just down to the `moduleDeps` ending up as class files rather than JARs, so `addDependencies` just ignores them.

# Result
I've added a fairly fragile heuristic to grab the `compile.dest/classes` paths from the `upstreamAssemblyClasspath` passed into `MDBuild` and use them with `addToClasspath`, which recursively copies the passed directories into `/app/classpath/` (preserving the directory structure) and adds `/app/classpath/` to the default entrypoint.

It would probably be better to use `transitiveLocalClasspath` since it _only_ contains `moduleDeps` outputs, but I figured I'd get this out sooner than later so I don't have to maintain my little workaround.

# Workaround

Until this is released, the same effect can be achieved with a line like

```
builder.addToClasspath(transitiveLocalClasspath().filter(_.path.last.endsWith("classes")).map(_.path.wrapped).toList.asJava
```

in an overridden `getJavaBuilder`.

### CODE REVIEW

File: mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala [hunk 1]
```diff
@@ -51,8 +51,12 @@ object MDBuild {
     val (upstreamClassSnapShot, upstreamClass) =
       buildSettings.upstreamAssemblyClasspath.partition(MDShared.isSnapshotDependency(_))
 
-    javaBuilder.addDependencies(upstreamClass.filter(x => os.exists(x.path)).map(x => x.path.wrapped).toList.asJava)
+    val (moduleDeps, libraryDeps) =
+      upstreamClass.partition(MDShared.isModuleDep(_))
+
+    javaBuilder.addDependencies(libraryDeps.filter(x => os.exists(x.path)).map(x => x.path.wrapped).toList.asJava)
     javaBuilder.addSnapshotDependencies(upstreamClassSnapShot.map(_.path.wrapped).toList.asJava)
+    javaBuilder.addToClasspath(moduleDeps.filter(x => os.isDir(path)).map(_.path.wrapped).toList.asJava)
```
Comment (sideeffffect) [inline mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala:59]:
```suggestion
    javaBuilder.addToClasspath(moduleDeps.filter(x => os.isDir(x.path)).map(_.path.wrapped).toList.asJava)
```
typo?

Comment (sideeffffect) [inline mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala:59]: There's also `javaBuilder.addProjectDependencies` which accepts JARs, if they're available.

Comment (sideeffffect) [inline mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala:59]: Note that `addToClasspath` expects individual files, not a directory.

File: mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala [hunk 2]
```diff
@@ -51,8 +51,12 @@ object MDBuild {
     val (upstreamClassSnapShot, upstreamClass) =
       buildSettings.upstreamAssemblyClasspath.partition(MDShared.isSnapshotDependency(_))
 
-    javaBuilder.addDependencies(upstreamClass.filter(x => os.exists(x.path)).map(x => x.path.wrapped).toList.asJava)
+    val (moduleDeps, libraryDeps) =
+      upstreamClass.partition(MDShared.isModuleDep(_))
+
+    javaBuilder.addDependencies(libraryDeps.filter(x => os.exists(x.path)).map(x => x.path.wrapped).toList.asJava)
```
Comment (sideeffffect) [inline mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala:57]: Also, why are there the `filter`s? Aren't they just hiding potential bugs? All the files must exist. If not, something somewhere is broken, no?

Comment (GeorgOfenbeck) [reply mill-docker/src/com/ofenbeck/mill/docker/MDBuild.scala:57]: @jhenahan thanks for the PR - generally the multi module build is definitly far from ideal yet (hence also part of the future work in the readme).

When i pushed the first release I tried to port a pre-existing project [libro-finito](https://github.com/LaurenceWarne/libro-finito) using the mill included docker module to this plugin -> and immediatly realized i havent covered multi module nicely.

Generally I would like to test / implement the multi module build in more detail - to catch more corner cases.
Is the repo you are using it in public? (didnt find it on first glance)

I have to admit right now I'm a bit swamped with Job Hunting/Interviews/Interview Prep (current Job Market sucks 🙈) - but once i have something signed i should be able to get back onto it and invest some time.
Just a heads up why things will not move much next month. I will ping / mention you when I'm back onto it 

@sideeffffect 
Thats by the way why there is an "exists" clause at the line you pointed out. You are absolutly right - its covering bugs - in particular in this multi module setup.
At the time it seemed to be issues with what mill reported for multi module builds and i did not want to get into the depth of it (as it ment digging deeper into the mill internals) and put it for "future work"  😅

**Comment by sideeffffect** [comment]
Hello everybody, is there something I can help with to move this forward?
It would be pity to have this bug go unfixed.

**Comment by jhenahan** [comment]
Sorry, got bogged down in some other things and wasn't working with Mill for a bit there. Let me take a look at the comments and see if I can polish this up a bit and then we can get it landed.

**Comment by sideeffffect** [comment]
@jhenahan @GeorgOfenbeck ping :smile_cat:

**Comment by sideeffffect** [comment]
Hello @jhenahan @GeorgOfenbeck , do you have any idea when you will be continuing the work on this PR?

**Comment by jhenahan** [comment]
Sorry, I rolled off the work project that motivated this in the first place (and haven't been on other mill projects in the interim), but I've got some bandwidth this week. I'll read back through the comments here and see if I can put something less completely hacky together to close this out.
